kbfsgit/runner.go

```diff
@@ -0,0 +1,287 @@
+	}
+}
+
+type runner struct {
+	config libkbfs.Config
+	log    logger.Logger
+	h      *libkbfs.TlfHandle
+	repo   string
+	uniqID string
+	input  io.Reader
+	output io.Writer
+}
+
+// newRunner creates a new runner for git commands.  It expects `repo`
+// to be in the form "keybase://private/user/reponame".
+func newRunner(ctx context.Context, config libkbfs.Config, repo string,
+	input io.Reader, output io.Writer) (*runner, error) {
+	tlfAndRepo := strings.TrimPrefix(repo, kbfsgitPrefix)
+	parts := strings.Split(tlfAndRepo, repoSplitter)
+	if len(parts) != 3 {
+		return nil, errors.Errorf("Repo should be in the format "+
+			"%s<tlfType>%s<tlf>%s<repo>, but got %s",
+			kbfsgitPrefix, repoSplitter, repoSplitter, tlfAndRepo)
+	}
+
+	var t tlf.Type
+	switch parts[0] {
+	case publicName:
+		t = tlf.Public
+	case privateName:
+		t = tlf.Private
+	case teamName:
+		t = tlf.SingleTeam
+	default:
+		return nil, errors.Errorf("Unrecognized TLF type: %s", parts[0])
+	}
+
+	h, err := getHandleFromFolderName(ctx, config, parts[1], t)
+	if err != nil {
+		return nil, err
+	}
+
+	// Use the device ID and PID to make a unique ID (for generating
+	// temp files in KBFS).
+	session, err := libkbfs.GetCurrentSessionIfPossible(
+		ctx, config.KBPKI(), h.Type() == tlf.Public)
+	if err != nil {
+		return nil, err
+	}
+	uniqID := session.VerifyingKey.String() + "-" + string(os.Getpid())
+
+	return &runner{
+		config: config,
+		log:    config.MakeLogger(""),
+		h:      h,
+		repo:   parts[2],
+		uniqID: uniqID,
+		input:  input,
+		output: output}, nil
+}
+
+func (r *runner) handleCapabilities() error {
+	caps := []string{
+		gitCmdFetch,
+		gitCmdPush,
+	}
+	for _, c := range caps {
+		_, err := r.output.Write([]byte(c + "\n"))
+		if err != nil {
+			return err
+		}
+	}
+	_, err := r.output.Write([]byte("\n"))
+	return err
+}
+
+func (r *runner) initRepoIfNeeded(ctx context.Context) (
+	*gogit.Repository, error) {
+	rootNode, _, err := r.config.KBFSOps().GetOrCreateRootNode(
+		ctx, r.h, libkbfs.MasterBranch)
+	if err != nil {
+		return nil, err
+	}
+
+	lookupOrCreateDir := func(n libkbfs.Node, name string) (
+		libkbfs.Node, error) {
+		newNode, _, err := r.config.KBFSOps().Lookup(ctx, n, name)
+		switch errors.Cause(err).(type) {
+		case libkbfs.NoSuchNameError:
+			newNode, _, err = r.config.KBFSOps().CreateDir(ctx, n, name)
+			if err != nil {
+				return nil, err
+			}
+		case nil:
+		default:
+			return nil, err
+		}
+		return newNode, nil
+	}
+
+	repoDir, err := lookupOrCreateDir(rootNode, kbfsRepoDir)
+	if err != nil {
+		return nil, err
+	}
+	_, err = lookupOrCreateDir(repoDir, r.repo)
+	if err != nil {
+		return nil, err
+	}
+
+	fs, err := libfs.NewFS(
+		ctx, r.config, r.h, path.Join(kbfsRepoDir, r.repo), r.uniqID)
+	if err != nil {
+		return nil, err
+	}
+
+	// We store the config in memory for two reasons. 1) gogit/gcfg
+	// has a bug where it can't handle backslashes in remote URLs, and
+	// 2) we don't want to flush the remotes since they'll contain
```

Comment:
In go-git, as far as I know, there is no way to hook two `go-git.Repository` instances together to do fetches/pulls between them.  One of the two repos has to be defined as a "remote" to the other one in order to use the nice `Fetch` and `Pull` commands.  (There might be other more involved ways to transfer objects manually one-by-one, but that seems like it would be pretty sad.)

Since there is no standard remote protocol for keybase yet (that's what we're building!), it's not supported by go-git itself.  That means our only option is to treat the local on-disk repo as a "remote" with respect to the bare KBFS repo, and do everything in reverse: for example, when a user does a push, we actually fetch from the local repo and write the objects into the bare repo.

It's confusing, but I haven't seen a way around this yet.  Let me know if you think of one.

FYI, the next PR in this series will be changing the storer to filter out the remotes, rather than store them in memory, since we still need some of the config file to be written to KBFS.  But these temporary remotes should never leave the local process, they're only for local convenience to work around go-git limitations.